File: CSES/Python/CountingRooms.py
from queue import Queue

dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]
n, m = map(int, input().split())
grid = [list(input()) for _ in range(n)]

def bfs(a, b):
    q = Queue()
    q.put((a, b))
    grid[a][b] = '#'
    while not q.empty():
        i, j = q.get()
        for dx, dy in dirs:
            x, y = i + dx, j + dy
            if 0 <= x < n and 0 <= y < m and grid[x][y] == '.':
                grid[x][y] = '#'
                q.put((x, y))

ans = 0
for i in range(n):
    for j in range(m):
        if grid[i][j] == '.':
            bfs(i, j)
            ans += 1
print(ans)

File: CSES/CountingRooms.cpp
#include <bits/stdc++.h>
using namespace std;

#define FAST_IO ios::sync_with_stdio(false); cin.tie(nullptr);

int n, m;
vector<string> grid;
vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

void bfs(int x, int y) {
    queue<pair<int, int>> q;
    q.push({x, y});
    grid[x][y] = '#';  

    while (!q.empty()) {
        auto [cx, cy] = q.front();
        q.pop();

        for (auto [dx, dy] : directions) {
            int nx = cx + dx, ny = cy + dy;
            if (nx >= 0 && nx < n && ny >= 0 && ny < m && grid[nx][ny] == '.') {
                grid[nx][ny] = '#';  
                q.push({nx, ny});
            }
        }
    }
}

int countRooms() {
    int roomCount = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (grid[i][j] == '.') {  
                bfs(i, j);
                roomCount++;
            }
        }
    }
    return roomCount;
}

int main() {
    FAST_IO;
    cin >> n >> m;
    grid.resize(n);
    
    for (int i = 0; i < n; i++) {
        cin >> grid[i];
    }

    cout << countRooms() << "\n";
    return 0;
}
